### PR TITLE
Ssong sumo 76740 metadata

### PR DESCRIPTION
**Pull request by samjsong**
- Added log-opt `tag` support. https://docs.docker.com/engine/admin/logging/log_tags/
- Added log-opt `sumo-source-category`. Currently supports explicit strings, but can specify `{{.Tag}}` in the string to include the tag in the _sourceCategory. For example: `test_category/{{.Tag}}/hello`
- Added HTTP headers for `X-Sumo-Name` and `X-Sumo-Category`.
- Updated README to reflect changes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,9 +37,10 @@ $ docker run --log-driver=sumologic --log-opt sumo-url=https://<deployment>.sumo
 ### Sumo Logic Options
 To specify additional logging driver options, you can use the `--log-opt NAME=VALUE` flag.
 
-| Option                      | Required? | Default Value | Description
-| --------------------------- | :-------: | :-----------: | -------------------------------------- |
+| Option                    | Required? | Default Value | Description
+| ------------------------- | :-------: | :-----------: | -------------------------------------- |
 | sumo-url                  | Yes       |               | HTTP Source URL
+| sumo-source-category      | No        | `dockerlog` | Source category to appear when searching on Sumo Logic by `_sourceCategory`. To include the log tag in the source category, use `{{.Tag}}`; see `tag`.
```
Comment (rvmiller89) [inline README.md]: we should add the same support for `_sourceHost` and `_sourceName`

Comment (samjsong) [reply README.md]: should we allow the tag to be in `_sourceHost` and `_sourceName` as well?

Comment (rvmiller89) [reply README.md]: yep, basically 3 ways for customer to specify the metadata

**Comment by rvmiller89** [comment]
I'm confused on the usage of `--log-opt tag`.  I would think we can actually skip specifying this, and that the user could specify the 3 metadata directly, such as
```
--log-opt sumo-source-category="Category_{{.ImageName}}" \
--log-opt sumo-source-name="{{.Name}}/{{.ID}}" \
```
etc.

**Comment by samjsong** [comment]
Yeah, I agree that would probably be simpler. One issue is that the tag option is a Docker supported thing: https://docs.docker.com/engine/admin/logging/log_tags/ so it makes sense to follow this convention of using the `"tag"` log-opt.

A smaller issue is that Docker provides in their code a way to easily parse the tag in their `loggerutils` package: `tag, err := loggerutils.ParseLogTag(info, loggerutils.DefaultTemplate)`. To support the functionality you mention, I would have to ignore this package (they hard-coded the key `"tag"`, so it wouldn't work with `"sumo-source-category"`), but if you feel that this would be a much better design, I can make that change

**Comment by rvmiller89** [comment]
Yea I think we'll want to look into this more, since we have 3 metadata fields, but only 1 `tag` field capable of template markup, which limits the usefulness of metadata.  Thoughts @bin3377 ?

**Comment by samjsong** [comment]
Definitely. One thing we could do is to support the tag in the backend, in which case one possible approach would be to support templating in the 3 metadata fields, and also to support a fourth `tag` metadata field.

**Comment by rvmiller89** [comment]
I wonder if we could:

1. examine all 3 metadata fields to find set of template keys
2. create an artificial tag string containing each key separated by separator, e.g. `{{.ID}}_{{. ImageName}}_{{.ImageName}}`
3. use `ParseLogTag ` to resolve the template, split again, and create a lookup map of each key to resolved value
4. use lookup map to string-replace the original metadata fields from (1)

**Comment by bin3377** [comment]
The problem here is: the native docker log/stats blades don't have ability to resolve the template in `-tag` (which is a go-lang feature). So for sending metadata in _sourceCategory (and other 2), I proposed to use a Sumo owned tokenized template as `{{.MARKUP}}` `{{#LABEL}}` and `{{$ENV}}`

So if we want make the docker logging driver has best compatible with native docker blades, Ideally `sumo-source-XXXX` should also support all 3 types of token and also add {{.Tag}} as one extra {{.MARKUP}} so customer can get benefit by both features (from docker -tag or from Sumo {{}})

And we can even give sumo-source-category a default value of `{{.Tag}}` so if customer just using `-tag` as in other logging driver, it will be sent as _sourceCategory

**Comment by bin3377** [comment]
@yuting-liu since we may want the native blades and logging driver has identical syntax on the metadata template

**Comment by bin3377** [comment]
@latkin  @rvmiller89 apply the change we discussed today. Please take a look before I merge it

**Comment by bin3377** [comment]
@latkin @rvmiller89 can you guys make it higher priority since @bgoleno want it out next week and we still have some validating process on the docker store side after we finalize the code

File: DEVELOPER.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# Guide for developers
+This project is a plugin working in docker engine for delivering log to Sumo Logic cloud service through HTTP source.
```
Comment (latkin) [inline DEVELOPER.md]: Fix some typos:

```
This project is a plugin for the docker engine, which delivers logs to Sumo Logic by pushing log messages through an HTTP source.
This is the guide for developers who want build and extend the plugin. If you just want use this plugin in your docker environment, please refer to the [readme file](README.md).
```

File: DEVELOPER.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  ```bash
+  $ go get -d ./...
+  ```
+
+## Build and install plugin to docker
+In bash, run:
+```bash
+$ sudo bash ./plugin_install.sh
+```
+If everything goes fine, you can verify the plugin correctly installed with:
```
Comment (latkin) [inline DEVELOPER.md]: `If everything goes fine, you can verify that the plugin is correctly installed with:`

File: DEVELOPER.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+$ sudo bash ./plugin_uninstall.sh
+```
+
+## Run sanity test
+  * Make sure the plugin is installed and enabled
+  * In bash, run:
+```bash
+$ docker run --log-driver=sumologic --log-opt sumo-url=<url> -i -t ubuntu bash
+```
+This will create a bash session in docker container and send all console contents to Sumo HTTP source as log lines
```
Comment (latkin) [inline DEVELOPER.md]: `This will create a bash session in a docker container and send all console contents to a Sumo Logic HTTP source as log lines`

File: DEVELOPER.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+## Run unit test
+The unit test is written in `XXX_test.go` which `XXX` is the module to be tested. You can launch all unite tests with:
```
Comment (latkin) [inline DEVELOPER.md]: unite = unit

File: README.md [hunk 6]
```diff
@@ -2,7 +2,7 @@
 
 A Docker logging driver plugin to send logs to Sumo Logic.
 
-**Disclaimer:** This plugin is still being developed.  We recommend using this plugin in non-production environments.
+**Disclaimer:** This plugin is still being developed. We recommend using this plugin in non-production environments.
```
Comment (latkin) [inline README.md]: If we are announcing this at Illuminate we should remove this warning. (can be separate PR)

File: README.md [hunk 7]
```diff
@@ -37,9 +37,12 @@ $ docker run --log-driver=sumologic --log-opt sumo-url=https://<deployment>.sumo
 ### Sumo Logic Options
 To specify additional logging driver options, you can use the `--log-opt NAME=VALUE` flag.
 
-| Option                      | Required? | Default Value | Description
-| --------------------------- | :-------: | :-----------: | -------------------------------------- |
+| Option                    | Required? | Default Value | Description
+| ------------------------- | :-------: | :-----------: | -------------------------------------- |
 | sumo-url                  | Yes       |               | HTTP Source URL
+| sumo-source-category      | No        |               | Source category to appear when searching on Sumo Logic by `_sourceCategory`. Use `{{Tag}}` as the placeholder of `tag` option. If not specified, the source category of HTTP source will be used
```
Comment (latkin) [inline README.md]: For all of these, "searching on" should be "searching in"

Comment (latkin) [inline README.md]: `If not specified, the default source category configured for the HTTP source will be used.`

File: README.md [hunk 8]
```diff
@@ -37,9 +37,12 @@ $ docker run --log-driver=sumologic --log-opt sumo-url=https://<deployment>.sumo
...
+| sumo-source-name          | No        |               | Source name to appear when searching on Sumo Logic by `_sourceName`. Use `{{Tag}}` as the placeholder of `tag` option. If not specified, it will be the container's name
```
Comment (latkin) [inline README.md]: "If not specified, it will be the container's name" This should be in the "Default Value" column it sounds like

Comment (latkin) [inline README.md]: "Use `{{Tag}}` as the placeholder of `tag` option." I think wording could be improved -- is it correct to say "Within the source name, the token `{{Tag}}` will be replaced with the value of the Docker `tag` option."

Similar for category/host.

File: README.md [hunk 9]
```diff
@@ -49,6 +52,9 @@ To specify additional logging driver options, you can use the `--log-opt NAME=VA
 | sumo-root-ca-path         | No        |               | Set the path to a custom root certificate.
 | sumo-server-name          | No        |               | Name used to validate the server certificate. By default, uses hostname of the `sumo-url`.
 | sumo-queue-size           | No        | 100           | The maximum number of log batches of size `sumo-batch-size` we can store in memory in the event of network failure, before we begin dropping batches. Thus in the worst case, the plugin will use `sumo-batch-size` * `sumo-queue-size` bytes of memory per container (default 100 MB).
+| tag                       | No        | {{.ID}}       | Specify a tag for messages, which interprets some markup. Default value is {{.ID}} (first 12 characters of the container ID). Refer to the [tag log-opt documentation] for customizing the log tag format.
```
Comment (latkin) [inline README.md]: Would update wording to "Specifies a tag for messages, which can be used in the 'source category', 'source name', and source host' fields. Certain tokens of the form `{{X}}` are supported. Default value is `{{.ID}}`, the first 12 characters of the container ID. Refer to the [tag log-opt documentation] for more information and a list of supported tokens."

File: driver.go [hunk 10]
```diff
@@ -191,6 +234,31 @@ func (sumoDriver *sumoDriver) StopLogging(file string) error {
   return nil
 }
 
+func interpretAll(re *regexp.Regexp, input string, dictionary map[string]string) string {
+  result := ""
+  lastIndex := 0
+
+  for _, v := range re.FindAllSubmatchIndex([]byte(input), -1) {
+      groups := []string{}
+      for i := 0; i < len(v); i += 2 {
```
Comment (latkin) [inline driver.go]: Why is it incrementing by 2?

File: README.md (hunk 8 above)
Comment (bin3377) [reply README.md]: it's a little bit different, since default value is always a fixed value but these 3 options are actually a behavior (since you cannot specify a `container's name` no matter whatever string you assigned here)

File: driver.go (hunk 10 above)
Comment (bin3377) [reply driver.go]: it's because the return of the FindAllSubmatchIndex() actually return a pattern as begin/end index of the matching sections

File: logger.go [hunk 11]
```diff
@@ -162,6 +162,11 @@ func (sumoLogger *sumoLogger) sendLogs(logs []*sumoLog) error {
   if sumoLogger.gzipCompression {
     request.Header.Add("Content-Encoding", "gzip")
   }
+  if sumoLogger.sourceCategory != "" {
```
Comment (latkin) [inline logger.go]: Does `!= ""` also handle `nil`/null?

File: README.md [hunk 12]
```diff
@@ -37,9 +37,12 @@ $ docker run --log-driver=sumologic --log-opt sumo-url=https://<deployment>.sumo
...
 
-| Option                      | Required? | Default Value | Description
-| --------------------------- | :-------: | :-----------: | -------------------------------------- |
+| Option                    | Required? | Default Value | Description
+| ------------------------- | :-------: | :-----------: | -------------------------------------- |
 | sumo-url                  | Yes       |               | HTTP Source URL
+| sumo-source-category      | No        |               | Source category to appear when searching on Sumo Logic by `_sourceCategory`. Use `{{Tag}}` as the placeholder of `tag` option. If not specified, the source category of HTTP source will be used
+| sumo-source-name          | No        |               | Source name to appear when searching on Sumo Logic by `_sourceName`. Use `{{Tag}}` as the placeholder of `tag` option. If not specified, it will be the container's name
+| sumo-source-host          | No        |               | Source host to appear when searching on Sumo Logic by `_sourceHost`. Use `{{Tag}}` as the placeholder of `tag` option. If not specified, it will be the machine host name
```
Comment (latkin) [inline README.md]: "If not specified, it will be the machine host name" should be in "default value" column

File: logger.go [hunk 13]
```diff
@@ -162,6 +162,11 @@ func (sumoLogger *sumoLogger) sendLogs(logs []*sumoLog) error {
...
+    request.Header.Add("X-Sumo-Category", sumoLogger.sourceCategory)
+  }
+  request.Header.Add("X-Sumo-Name", sumoLogger.sourceName)
```
Comment (latkin) [inline logger.go]: I know these should never be empty, but would be good to do `!= ""` check on these too.

Comment (bin3377) [reply logger.go]: ok. I can add the checking

File: logger.go (hunk 11 above)
Comment (bin3377) [reply logger.go]: That's because in golang the string cannot be nil (but `*string` can)

**Comment by latkin** [comment]
Looking at https://docs.docker.com/engine/admin/logging/log_tags/ it seems that if users specify a `tag` logopt, then the value of the `tag` will be injected into every log message? Is this the case?

I thought we were using it simply as an easy way to do token evaluation, then only expected the result to be included in sourceCategory, sourceHost, etc.

File: README.md (hunk 8 above)
Comment (latkin) [reply README.md]: I woudl suggest using `code style` for the fixed value ones, and regular text description for others.

From casual user's perspective, if the "Default Value" column is empty, then they would expect no default value.

**Comment by bin3377** [comment]
The behavior of tag is decided by the logging driver. For example in [Splunk plugin](https://docs.docker.com/engine/admin/logging/splunk/#message-formats) The tag can become a json field or a prefix depends on the format of message

**Comment by latkin** [comment]
Ok, so when a Docker + Sumo user specifies a `tag`, the tag value will NOT be injected into log lines? Because it is up to our driver to decide what to do with it?

File: README.md (hunk 8 above)
Comment (bin3377) [reply README.md]: ok. make sense

**Comment by latkin** [comment]
LGTM

**Comment by bin3377** [comment]
Merge to make a tag. feel free to comment

**Comment by aidansteele** [comment]
Did you end up implementing the `{{.MARKUP}}`, `{{#LABEL}}` and `{{$ENV}}` functionality @bin3377? I would like to be able to use Docker container labels in both the source host and source category, but right now this appears to not be possible with the current tag workaround.

**Comment by bin3377** [comment]
Not yet. File an issue https://github.com/SumoLogic/sumologic-docker-logging-driver/issues/21 for tracking it. But since we will be consistent with the tag using in other SumoLogic source types, the tags will probably be changed to the formats described on [this page](https://help.sumologic.com/Send-Data/Applications-and-Other-Data-Sources/Docker/01-Collect-Events-and-Statistics-for-the-Docker-App)
